### PR TITLE
Patch optimizer for adding param_groups

### PR DESCRIPTION
**Pull request by Bonytu**
### Description ###
Patch optimizer for adding param groups.
#### Test Options ####
  - [ ] fast test
  - [ ] full test - HPO
  - [ ] full test - NAS
  - [ ] full test - compression

### Checklist ###
  - [ ] test case
  - [ ] doc

### CODE REVIEW

File: nni/contrib/compression/utils/evaluator.py [hunk 1]
```diff
@@ -418,9 +464,40 @@ def unbind_model(self):
         else:
             _logger.warning('Did not bind any model, no need to unbind model.')
 
-    def _patch_configure_optimizers(self):
+    def patch_optim_param_group(self, module_name_param_dict: Dict[str, List[Tensor]] | None = None):
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:467]: why `module_name_param_dict` can be None?

File: nni/contrib/compression/utils/evaluator.py [hunk 2]
```diff
@@ -185,6 +186,51 @@ def unbind_model(self):
...
+
+                for name in name_lis:
+                    if module_name in name:
+                        return param_group
+
+            return None
+
+        def add_param(param_lis: List[Tensor], target_param_group: Dict, optimizer: Optimizer):
+            for param in param_lis:
+                new_param_group = target_param_group.copy()
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:212]: copy() is a shadow copy, is this enough? or will some other key-value pairs be related to `params`?

Comment (Bonytu) [reply nni/contrib/compression/utils/evaluator.py:212]: after check optim algorithms in the torch.optim, I find that types of all values in the param_group are  Union[str, float, bool, Tuple[float, float]], so I think copy() is enough.

File: nni/contrib/compression/utils/evaluator.py (hunk 1 above)
Comment (Bonytu) [reply nni/contrib/compression/utils/evaluator.py]: done

File: nni/contrib/compression/utils/evaluator.py [hunk 3]
```diff
@@ -225,14 +235,14 @@ def add_param(param_lis: List[Tensor], target_param_group: Dict, optimizer: Opti
             is_find_param_group = False
             for optimizer in optimizers:
                 param_groups = optimizer.param_groups
-                target_param_group = find_param_group(param_groups, module_name)
-                if target_param_group is not None:
+                target_param_group_idx = find_param_group(param_groups, module_name)
+                if target_param_group_idx > 0:
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:233]: `= 0`?

File: nni/contrib/compression/utils/evaluator.py [hunk 4]
```diff
@@ -185,6 +186,63 @@ def unbind_model(self):
...
+                        return i
+
+            return -1
+
+        def add_param(param_lis: List[Tensor], target_param_group_idx: int, optimizer: Optimizer):
+            assert target_param_group_idx < len(optimizer.param_groups)
+            target_param_group = optimizer.param_groups[target_param_group_idx]
+            for param in param_lis:
+                # copyed from torch
+                if not isinstance(param, toech.Tensor):
```
Comment (J-shang) [inline nni/contrib/compression/utils/evaluator.py:218]: typo `torch`

**Review by J-shang — APPROVED** [review]

**Review by QuanluZhang — APPROVED** [review]

**Pull request merged**